### PR TITLE
Add wait_period config before abandoning executions during shutdown.

### PR DESCRIPTION
**Pull request by khushboobhatia01**
This pull request adds a new `abandon_wait_config` which will allow user to configure wait period before worker starts abandoning incomplete executions during shutdown. This will ensure short lived executions can complete instead of just abandoning.

Why is this change needed?
- In K8S HA environment we have a `terminationGracePeriod` config which waits for `x` seconds before killing the pod.
We plan to set this config, so that short lived executions can complete without any issue. With the current implementation actionrunner will directly start abandoning incomplete executions and `terminationGracePeriod` config won't be of much use.

### CODE REVIEW

File: st2actions/st2actions/worker.py [hunk 1]
```diff
@@ -134,6 +137,13 @@ def process(self, liveaction):
 
     def shutdown(self):
         super(ActionExecutionDispatcher, self).shutdown()
+        abandon_wait_period = cfg.CONF.actionrunner.abandon_wait_period
+        LOG.debug(
```
[the remark this answers is not in the document]
Comment (khushboobhatia01) [reply st2actions/st2actions/worker.py]: Addressed

File: st2common/st2common/config.py [hunk 2]
```diff
@@ -477,6 +477,13 @@ def register_opts(ignore_errors=False):
                 "that size"
             ),
         ),
+        cfg.IntOpt(
+            "abandon_wait_period",
+            default=0,
+            help=(
+                "Wait period in seconds after which actionrunner starts abandoning incomplete executions during shutdown."
```
[the remark this answers is not in the document]
Comment (khushboobhatia01) [reply st2common/st2common/config.py:484]: Addressed

**Comment by khushboobhatia01** [comment]
> Thanks for the contribution.
> 
> The change indeed looks reasonable, but perhaps also a corresponding documentation PR may be in order - aka in what kind of situations users may want to set this value to something larger than 0 (basically what you described as your motivation for the change - in case users have short running executions or similar).

Thanks @Kami for the review. I've added the new config details in the CHANGELOG.

**Comment by amanda11** [comment]
@khushboobhatia01 / @m4dcoder Given the discussions about a clean shutdown behaviour, is this PR still targetted for the 3.6.0 release.
Are we aiming to put in this as a workaround or wait until we get a full solution instead?

If we are going to wait, then we should probably remove it from the 3.6 milestone and project.

**Comment by cognifloyd** [comment]
I think a pause is good, but only
- if it stops consuming new actions first, and
- if there there is something running to wait for.

So, I don't think this should go in 3.6

**Comment by khushboobhatia01** [comment]
> I think a pause is good, but only
> 
> * if it stops consuming new actions first, and
> * if there there is something running to wait for.
> 
> So, I don't think this should go in 3.6

Agree. 
Current implementation just abandons all ongoing executions which is not graceful.

We could think of a similar approach as discussed for workflow engine graceful shutdown https://github.com/StackStorm/st2/discussions/5373)

- On shutdown, the worker should remove itself from the service registry.
- If there are other workers in the service registry, then do the shutdown sequence. The shutdown sequence should be something like stop accepting incoming requests and wait until this worker completes processing any requests.
- Else if there are no other worker in the service registry, then the worker needs to stop accepting requests, pause all running action executions, and then proceed with rest of the shutdown sequence.
- On worker startup, any action executions paused by the shutdown should be resumed.

In this case abandonment will happen only for orphaned action executions stuck in some running/canceling/pausing state

Thoughts @cognifloyd  @m4dcoder?

**Comment by khushboobhatia01** [comment]
> @khushboobhatia01 / @m4dcoder Given the discussions about a clean shutdown behaviour (#5367, is this PR still targetted for the 3.6.0 release. Are we aiming to put in this as a workaround or wait until we get a full solution instead?
> 
> If we are going to wait, then we should probably remove it from the 3.6 milestone and project. If we want to put this in to improve the situation, then let us know if review is required.

@amanda11 I think this PR is still open for discussion. I had created this MR before the workflow engine graceful shutdown discussion.

**Comment by cognifloyd** [comment]
Yeah. We should use use the same approach for workflow engine and actionrunner.

As an interim PR, I think getting the services to add/remove self from registry would be a good start. Everything at once would probably be more difficult to review & merge.

**Comment by khushboobhatia01** [comment]
> Yeah. We should use use the same approach for workflow engine and actionrunner.
> 
> As an interim PR, I think getting the services to add/remove self from registry would be a good start. Everything at once would probably be more difficult to review & merge.

Yes, will start with that.

**Pull request closed without merging**